numba/withcontexts.py

```diff
@@ -0,0 +1,184 @@
+        body_block: sequence[int]
+            A sequence of int's representing labels of the with-body
+        dispatcher_factory : callable
+            A callable that takes a `FunctionIR` and returns a `Dispatcher`.
+        """
+        raise NotImplementedError
+
+
+@typeof_impl.register(WithContext)
+def typeof_contextmanager(val, c):
+    return types.ContextManager(val)
+
+
+def _get_var_parent(name):
+    """Get parent of the variable given its name
+    """
+    # If not a temprary variable
+    if not name.startswith('$'):
+        # Return the base component of the name
+        return name.split('.', )[0]
+
+
+def _clear_blocks(blocks, to_clear):
+    """Remove keys in *to_clear* from *blocks*.
+    """
+    for b in to_clear:
+        del blocks[b]
+
+
+class _ByPassContextType(WithContext):
+    """A simple context-manager that tells the compiler to bypass the body
+    of the with-block.
+    """
+    def mutate_with_body(self, func_ir, blocks, blk_start, blk_end,
+                         body_blocks, dispatcher_factory):
+        # Determine variables that need forwarding
+        vlt = func_ir.variable_lifetime
+        inmap = {_get_var_parent(k): k for k in vlt.livemap[blk_start]}
+        outmap = {_get_var_parent(k): k for k in vlt.livemap[blk_end]}
+        forwardvars = {inmap[k]: outmap[k] for k in filter(bool, outmap)}
+        # Transform the block
+        _bypass_with_context(blocks, blk_start, blk_end, forwardvars)
+        _clear_blocks(blocks, body_blocks)
+
+
+bypass_context = _ByPassContextType()
+
+
+class _CallContextType(WithContext):
+    """A simple context-manager that tells the compiler lift the body of the
+    with-block as another function.
+    """
+    def mutate_with_body(self, func_ir, blocks, blk_start, blk_end,
+                         body_blocks, dispatcher_factory):
+        vlt = func_ir.variable_lifetime
+        inputs = vlt.livemap[blk_start]
+        outputs = vlt.livemap[blk_end]
+
+        lifted_blks = {k: blocks[k] for k in body_blocks}
+        _mutate_with_block_callee(lifted_blks, blk_start, blk_end,
+                                  inputs, outputs)
+
+        # XXX: transform body-blocks to return the output variables
+        lifted_ir = func_ir.derive(
+            blocks=lifted_blks,
+            arg_names=tuple(inputs),
+            arg_count=len(inputs),
+            force_non_generator=True,
+            )
+
+        dispatcher = dispatcher_factory(lifted_ir)
+
+        newblk = _mutate_with_block_caller(
+            dispatcher, blocks, blk_start, blk_end, inputs, outputs,
+            )
+
+        blocks[blk_start] = newblk
+        _clear_blocks(blocks, body_blocks)
+        return dispatcher
+
+
+call_context = _CallContextType()
+
+
+def _bypass_with_context(blocks, blk_start, blk_end, forwardvars):
+    """Given the starting and ending block of the with-context,
+    replaces a new head block that jumps to the end.
```

Comment:
Should this be `replaces the head block with a new block that jumps to the end`.?